pkg/odo/cli/dev/dev.go

```diff
@@ -268,3 +310,105 @@ It forwards endpoints with any exposure values ('public', 'internal' or 'none')
+		}
+	}
+
+	// 	Check that all container ports are valid and present in the Devfile
+portLoop:
+	for _, fPort := range forwardedPorts {
+		if fPort.ContainerName != "" {
+			if containerEndpoints, ok := containerEndpointMapping[fPort.ContainerName]; ok {
+				for _, endpoint := range containerEndpoints {
+					if endpoint.TargetPort == fPort.ContainerPort {
+						klog.V(1).Infof("container port %d matches %s endpoints of container %q", fPort.ContainerPort, endpoint.Name, fPort.ContainerName)
+						continue portLoop
+					}
+				}
+				errors = append(errors, fmt.Sprintf("container port %d does not match any endpoints of container %q in the devfile", fPort.ContainerPort, fPort.ContainerName))
+			} else {
+				errors = append(errors, fmt.Sprintf("container %q not found in the devfile", fPort.ContainerName))
+			}
+		} else {
+			// Validate that a custom port mapping for port forwarding config without a container targets a unique containerPort
+			if containers, ok := portContainerMapping[fPort.ContainerPort]; ok && len(containers) > 1 {
+				// this workaround makes sure the unit tests do not fail because of a change in the order
+				sort.Strings(containers)
+				msg := fmt.Sprintf("multiple container components (%s) found with same container port %d in the devfile, port forwarding must be defined with format <localPort>:<containerName>:<containerPort>", strings.Join(containers, ", "), fPort.ContainerPort)
+				if !strings.Contains(strings.Join(errors, ", "), msg) {
+					errors = append(errors, msg)
+				}
+
+				continue portLoop
+			}
+			for _, containerEndpoints := range containerEndpointMapping {
+				for _, endpoint := range containerEndpoints {
+					if endpoint.TargetPort == fPort.ContainerPort {
+						klog.V(1).Infof("container port %d matches %s endpoints of container %s", fPort.ContainerPort, endpoint.Name, fPort.ContainerName)
+						continue portLoop
+					}
+				}
+			}
+			errors = append(errors, fmt.Sprintf("container port %d not found in the devfile container endpoints", fPort.ContainerPort))
+		}
+	}
+	if len(errors) != 0 {
+		return errors, fmt.Errorf("values for --port-forward flag are invalid")
+	}
+	return nil, nil
+}
+
+// parsePortForwardFlag parses custom port forwarding configuration; acceptable patterns: <localPort>:<containerPort>, <localPort>:<containerName>:<containerPort>
+func parsePortForwardFlag(portForwardFlag []string) (forwardedPorts []api.ForwardedPort, err error) {
+	// acceptable examples: 8000:runtime_123:8080, 8000:9000, 8000:runtime:8080, 20001:20000
+	// unacceptable examples: :8000, 80000:
+	pattern := `^(\d{1,5})(:\w*)?:(\d{1,5})$`
+	portReg := regexp.MustCompile(pattern)
+	const largestPortValue = 65535
+	for _, portData := range portForwardFlag {
+		if !portReg.MatchString(portData) {
+			return nil, errors.New("ports are not defined properly, acceptable formats are: <localPort>:<containerPort>, <localPort>:<containerName>:<containerPort>")
```

Comment:
Suggestion to update the error message. In the example below, I intentionally passed a high port respecting the format, but the error message does not help identify the cause of the error:

```
$ odo dev --debug --port-forward 8089:3000 --port-forward 100001:runtime:5858
 ✗  ports are not defined properly, acceptable formats are: <localPort>:<containerPort>, <localPort>:<containerName>:<containerPort>
```

```suggestion
			return nil, errors.New("ports are not defined properly, acceptable formats are: <localPort>:<containerPort>, <localPort>:<containerName>:<containerPort>", where ports must be numbers in the range [1, 65535])
```